rust/arrow/src/array.rs

```diff
@@ -1659,13 +1710,80 @@ mod tests {
         field_types.push(Field::new("a", DataType::Boolean, false));
         field_types.push(Field::new("b", DataType::Int64, false));
         let struct_array_data = ArrayData::builder(DataType::Struct(field_types))
+            .len(4)
+            .add_child_data(boolean_data.clone())
+            .add_child_data(int_data.clone())
+            .build();
+        let struct_array = StructArray::from(struct_array_data);
+
+        assert_eq!(boolean_data, struct_array.column(0).data());
+        assert_eq!(int_data, struct_array.column(1).data());
+    }
+
+    #[test]
+    fn test_struct_array_slice() {
+        let mut null_bits: [u8; 1] = [0; 1];
+        bit_util::set_bit(&mut null_bits, 0);
+        bit_util::set_bit(&mut null_bits, 2);
+        bit_util::set_bit(&mut null_bits, 4);
+        let null_bit_buffer = Buffer::from(null_bits);
+
+        let boolean_data = ArrayData::builder(DataType::Boolean)
+            .len(5)
+            .add_buffer(Buffer::from(
+                [false, false, false, true, false].to_byte_slice(),
+            ))
+            .null_bit_buffer(null_bit_buffer.clone())
+            .build();
+        let int_data = ArrayData::builder(DataType::Int32)
+            .len(5)
+            .add_buffer(Buffer::from([42, 28, 19, 31, 54].to_byte_slice()))
+            .null_bit_buffer(null_bit_buffer.clone())
+            .build();
+
+        let mut field_types = vec![];
+        field_types.push(Field::new("a", DataType::Boolean, false));
+        field_types.push(Field::new("b", DataType::Int32, false));
+        let struct_array_data = ArrayData::builder(DataType::Struct(field_types))
+            .len(5)
             .add_child_data(boolean_data.clone())
             .add_child_data(int_data.clone())
+            .null_bit_buffer(null_bit_buffer)
```

Comment:
as an observation, we are sometimes prone to adding a bit buffer without nulls, or the inverse when using `ArrayData::builder`. Should I open a JIRA for checking that if `null_count > 0` when calling build, the bit_buffer is also set?
I think this might catch subtle errors that we make in future.

Reply:
Can you elaborate on this? If `null_count` is not specified in the input, it is calculated from `bit_buffer` so they should be consistent. Otherwise we'll just use `null_count`.

Reply:
Great point, I didn't inspect the function closely, my apologies for that. 
I was trying to say that nothing stops the user from doing the below:

```rust
ArrayData::builder(dt)
  .len(n)
  .null_count(Some(3)) // only setting null count, but not supplying bitmap
  .buffers(vec![my_buffer])
  .build();
```

The inverse where I supply just the bitmap isn't an issue because the null_count is calculated like you mention. It's the above case where I'm unsure of what should happen.

Reply:
Yes in this case it won't verify whether the bit buffer is consistent with null count. It will use the value from null count for `null_count()` but bit buffer for `is_valid()` and `is_null()`. 

With that said, I think the purpose of the null count parameter is to avoid calculating it from the bit buffer. To do the extra step of validating it seems to defeat that purpose.

Reply:
Okay, I understand. I've seen that in cpp and JS the null count is often left as unknown (-1).